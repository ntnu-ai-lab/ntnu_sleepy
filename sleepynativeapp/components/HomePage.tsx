import React from "react";
import { View, Text } from "react-native";
import { NavBar } from "./material/NavBar";
import { PageTemplate } from "./PageTemplate";

export function HomePage() {

    return (
        <PageTemplate>
            <View>
                <Text>
                    This is the homepage
                </Text>
            </View>
        </PageTemplate>
    )
}