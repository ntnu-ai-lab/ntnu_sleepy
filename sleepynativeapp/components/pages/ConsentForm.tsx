import React from "react";
import { View, Text, StyleSheet, ScrollView } from "react-native";
import { IconButton } from "../material/IconButton";
import { PageTemplate } from "../material/PageTemplate";
import Close from "../../assets/close.svg";
import { useNavigation } from "@react-navigation/native";

export function ConsentForm() {
  const navigation = useNavigation();

  const style = StyleSheet.create({
    sectionheading: {
        fontSize: 18,
        textTransform: "uppercase",
        marginVertical: 3
     },
    mainheading: {
        fontSize: 24,
        marginVertical: 10,
        textTransform: "uppercase"
    },
    bodytext: {
        marginVertical: 3
    },
    closeButton: {
        position: "absolute",
        width: "100%",
        top: 0,
        alignItems: "flex-end",
        paddingRight: 10,
        justifyContent: "center",
        paddingTop: 30,
        height: 80
    },
    pagewrapper: {
        padding: 10,
        marginTop: 70
    },
  });

  return (
    <View>
        <View style={style.closeButton}>
          <IconButton onClick={navigation.goBack}>
            <Close />
          </IconButton>
        </View>
      <ScrollView style={style.pagewrapper}>
        <View>
          <Text style={style.sectionheading}>
            Forespørsel om deltakelse i forskningsprosjektet
          </Text>
        </View>
        <View>
          <Text style={style.mainheading}>
            Oppfølging av pasienter med SØVNPROBLEMER via app: EN randomisert
            kontrollert studie
          </Text>
        </View>
        <Text style={style.bodytext}>
          Dette er et spørsmål til deg om å delta i et forskningsprosjekt for å
          undersøke nytten og brukervennligheten av en app («??») for personer
          med søvnproblemer. Til studien ønsker vi å rekruttere personer med
          kroniske søvnproblemer. Institutt for samfunnsmedisin og sykepleie ved
          Norges Teknisk-Naturvitenskapelige Universitet (NTNU) er ansvarlig for
          studien.
        </Text>
        <View>
          <Text style={style.sectionheading}>Hva innebærer PROSJEKTET?</Text>
        </View>
        <Text style={style.bodytext}>
          Vi vil be deg fylle inn et spørreskjema der vi innhenter opplysninger
          om din bakgrunn og helsetilstand – dette tar ca. 20 min. Etter at du
          har fylt inn spørreskjemaet vil du ved tilfeldig trekning bli fordelt
          til en kontrollgruppe eller en forsøksgruppe. Hvis du fordeles til
          kontrollgruppen vil du ikke få noe videre oppfølging fra oss og du vil
          ikke få tilgang til å bruke appen (se beskrivelse nedenfor). Deltakere
          i både forsøksgruppen og kontrollgruppen vil bli bedt om å svare på et
          spørreskjema etter 6 uker, 3, 6 og 9 måneder. Dette tar 15-25 min hver
          gang.{" "}
        </Text>
        <Text style={style.bodytext}>
          Hvis du fordeles til forsøksgruppen vil vi be deg om å bruke «??»
          appen i inntil 9 måneder. Appen vil gi deg råd om hvordan du bør
          håndtere søvnproblemene dine. Noen av rådene er basert på de
          opplysningene du oppgir i spørreskjemaet på starten av studien. I
          tillegg vil du via appen få ukentlige spørsmål om hvordan det går med
          plagene dine – dette tar ca. 2-3 min per uke. Du kan også bli invitert
          til et intervju der vi spør mer detaljert om hvordan du har brukt
          appen og hva som fungerte bra/mindre bra. Deltakelse i intervjuet er
          frivillig – dette kan gjennomføres på telefon.
        </Text>
        <Text style={style.bodytext}>
          I prosjektet vil vi innhente og registrere opplysninger om deg. Dette
          gjelder din e-post adresse og ditt telefonnummer – dette trenger vi
          får å kunne gi deg tilgang til de elektroniske spørreskjemaene som
          brukes i studien og for å kunne kontakte deg underveis i studien. I
          spørreskjemaene spør vi om helsemessige forhold som kan ha betydning
          for søvnproblemene dine.{" "}
        </Text>
        <View>
          <Text style={style.sectionheading}>Mulige fordeler og ulemper</Text>
        </View>
        <Text style={style.bodytext}>
          Hvis du fordeles til forsøksgruppen vil du få råd via appen om hvordan
          du bør håndtere søvnproblemene dine. Behandlingen består blant annet
          av informasjon om søvnproblemer, utfylling av søvndagbok, vaneendring
          og stimuluskontroll, søvnrestriksjon, og diverse mentale strategier.
          Uansett om du fordeles til forsøksgruppen eller kontrollgruppen vil du
          være med på å gi et viktig bidrag til utviklingen et nytt verktøy som
          kan være til stor hjelp for mange personer med søvnproblemer.
        </Text>
        <Text style={style.bodytext}>
          Det er ingen kjent risiko med deltakelse i studien, men
          søvnrestriksjon kan for enkelte være krevende i starten av
          behandlingen. Hvis du har spørsmål underveis i studien har du mulighet
          for å kontakte oss på telefon eller via e-post. Deltakelse i studien
          vil ikke ha noen innvirkning eller konsekvens for annen behandling du
          mottar.
        </Text>
        <View>
          <Text style={style.sectionheading}>
            Frivillig deltakelse og mulighet for å trekke sitt samtykke
          </Text>
        </View>
        <Text style={style.bodytext}>
          Det er frivillig å delta i prosjektet. Dersom du ønsker å delta,
          samtykker du ved registrering. Du kan når som helst og uten å oppgi
          noen grunn trekke ditt samtykke. Dette vil ikke få konsekvenser for
          din videre behandling (dvs. hvis du mottar behandling underveis i
          studien). Dersom du trekker deg fra prosjektet, kan du kreve å få
          slettet innsamlede opplysninger, med mindre opplysningene allerede er
          inngått i analyser eller brukt i vitenskapelige publikasjoner.
        </Text>
        <View>
          <Text style={style.sectionheading}>
            Hva skjer med opplysningene om deg?
          </Text>
        </View>
        <Text style={style.bodytext}>
          Informasjonen som registreres om deg skal kun brukes slik som
          beskrevet i hensikten med studien. Du har rett til innsyn i hvilke
          opplysninger som er registrert om deg og rett til å få korrigert
          eventuelle feil i de opplysningene som er registrert.
        </Text>
        <Text style={style.bodytext}>
          Alle opplysningene vil bli behandlet uten navn og fødselsnummer eller
          andre direkte gjenkjennende opplysninger. En kode knytter deg til dine
          opplysninger gjennom en navneliste.
        </Text>
        <Text style={style.bodytext}>
          Prosjektleder har ansvar for den daglige driften av
          forskningsprosjektet og at opplysninger om deg blir behandlet på en
          sikker måte. Informasjon om deg vil bli anonymisert etter
          prosjektslutt. Anonymiserte data vil lagres i inntil 30 år for å kunne
          brukes i videre forskning. Dataene som samles i prosjektet kan bidra
          til å utvikle en datamodell som i fremtiden kan inngå i en mulig
          kommersiell versjon av appen.
        </Text>
        <View>
          <Text style={style.sectionheading}>Forsikring </Text>
        </View>
        <Text style={style.bodytext}>
          Deltakelse i studien innebærer ingen kjent risiko for skade. Det er
          ikke tegnet egen forsikring for prosjektet, men NTNU har et generelt
          erstatningsansvar for denne type prosjekter hvis skade skulle oppstå.
        </Text>
        <View>
          <Text style={style.sectionheading}>Økonomi</Text>
        </View>
        <Text style={style.bodytext}>
          Alle som deltar i studien er med trekningen av en iPad (verdi ca. kr
          3000,- per stk). Utover dette vil du ikke motta noen godtgjørelse for
          deltakelse i studien.
        </Text>
        <View>
          <Text style={style.sectionheading}>Godkjenning</Text>
        </View>
        <Text style={style.bodytext}>
          Regional komité for medisinsk og helsefaglig forskningsetikk har
          vurdert prosjektet, og har gitt forhåndsgodkjenning (REK ###).
        </Text>
        <Text style={style.bodytext}>
          Etter ny personopplysningslov har dataansvarlig NTNU og prosjektleder
          Eivind Schjelderup Skarpsno et selvstendig ansvar for å sikre at
          behandlingen av dine opplysninger har et lovlig grunnlag. Dette
          prosjektet har rettslig grunnlag i EUs personvernforordning artikkel
          6a og artikkel 9 nr. 2 og ditt samtykke.
        </Text>
        <Text style={style.bodytext}>
          Du har rett til å klage på behandlingen av dine opplysninger til
          Datatilsynet.
        </Text>
        <View>
          <Text style={style.sectionheading}>kONTAKTOPPLYSNINGER</Text>
        </View>
        <Text style={style.bodytext}>
          Dersom du har spørsmål til prosjektet kan du ta kontakt med
          prosjektleder Eivind Schjelderup Skarpsno, tlf 97521297,
          eivind.s.skarpsno@ntnu.no.
        </Text>
        <View style={{height: 100}}/>
      </ScrollView>
      </View>
  );
}
