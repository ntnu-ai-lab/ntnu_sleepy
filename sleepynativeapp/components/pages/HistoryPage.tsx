import { StackActions, useNavigation } from "@react-navigation/native";
import { createNativeStackNavigator } from "@react-navigation/native-stack";
import React, { useEffect, useRef, useState } from "react";
import { View, Text } from "react-native";
import { useRecoilState } from "recoil";
import { getAllModules } from "../../api/modulesApi";
import { cachedModules } from "../../state/atoms";
import { Module } from "../../types/modules";
import { Alert } from "../material/Alert";
import { Button } from "../material/Button";
import { Card } from "../material/Card";
import { PageTemplate } from "../material/PageTemplate";
import { ModulePartPage } from "../module/ModulePartPage";

export function HistoryPage() {
  const navigation = useNavigation();
  const [thisModules] = useRecoilState(cachedModules);
  const [openAlert, setOpenAlert] = useState(false);

  useEffect(() => {
    console.log(thisModules);
  }, []);

  return (
    <PageTemplate>
      <View>
        {thisModules?.map((m, i) => {
          return (
            <Card key={m.id}>
              <Text>{m.title}</Text>
              <Button
                onClick={() => {
                  //@ts-ignore
                  navigation.navigate("home", {
                    screen: "History",
                    params: {
                      screen: "modulePart",
                      params: { part: m.parts[0] },
                    },
                  });
                }}
              >
                <Text>Se modul</Text>
              </Button>
            </Card>
          );
        })}
      </View>

      <View style={{ height: 80, width: "100%" }} />

      <Alert
        type={"alert"}
        content={
          "Klarte ikke 책 hente test, dette kan skyldes mangel p책 internett, eller mangel p책 kompetanse fra v책r side"
        }
        open={openAlert}
        setOpen={setOpenAlert}
      />
    </PageTemplate>
  );
}

export function HistoryNavigation() {
  const Stack = createNativeStackNavigator();

  return (
    <View style={{ height: "100%" }}>
      <Stack.Navigator
        screenOptions={{
          headerShown: false,
          headerTransparent: true,
          statusBarTranslucent: true,
          statusBarColor: "transparent",
          animation: "none",
        }}
        initialRouteName="moduleHistory"
      >
        <Stack.Screen name="moduleHistory" component={HistoryPage} />
        <Stack.Screen name="modulePart" component={ModulePartPage} />
      </Stack.Navigator>
    </View>
  );
}
