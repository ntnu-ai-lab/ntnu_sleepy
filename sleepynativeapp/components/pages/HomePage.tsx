import React from "react";
import { View, Text } from "react-native";
import { PageTemplate } from "../material/PageTemplate";

export function HomePage() {

    return (
        <PageTemplate>
            <View>
                <Text>
                    This is the homepage
                </Text>
            </View>
        </PageTemplate>
    )
}