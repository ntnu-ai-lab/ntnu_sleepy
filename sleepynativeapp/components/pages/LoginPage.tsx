import {
  SelfServiceLoginFlow,
  SubmitSelfServiceLoginFlowBody,
  SubmitSelfServiceLoginFlowWithPasswordMethodBody,
} from "@ory/kratos-client";
import {
  useFocusEffect,
  useNavigation,
  useRoute,
} from "@react-navigation/native";
import React, { useContext, useEffect, useState } from "react";
import { View, Dimensions, Text } from "react-native";
import { SessionContext } from "../../auth/Auth";
import { AuthContext } from "../../auth/AuthProvider";
import { handleFormSubmitError } from "../../auth/form";
import { ProjectContext } from "../../auth/ProjectProvider";
import { newKratosSdk } from "../../auth/Sdk";
import { colors } from "../../styles/styles";
import { Button } from "../material/Button";
import { Card } from "../material/Card";
import { PageTemplate } from "../material/PageTemplate";
import { TextField } from "../material/TextField";

export function LoginPage() {
  const [email, setEmail] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const navigation = useNavigation();
  const { project } = useContext(ProjectContext);
  const { setSession, session, sessionToken, isAuthenticated } =
    useContext(AuthContext);
  const [flow, setFlow] = useState<SelfServiceLoginFlow | undefined>(undefined);
  const route = useRoute();

  const initializeFlow = () =>
    newKratosSdk(project)
      .initializeSelfServiceLoginFlowWithoutBrowser(
        true,
        undefined,
        sessionToken
      )
      .then((response) => {
        const { data: flow } = response;
        // The flow was initialized successfully, let's set the form data:
        setFlow(flow);
      })
      .catch(console.error);

  useFocusEffect(
    React.useCallback(() => {
      initializeFlow();

      return () => {
        setFlow(undefined);
      };
    }, [project])
  );

  useEffect(() => {
    if (isAuthenticated && sessionToken && session) {
      //@ts-ignore
      navigation.navigate("home");
    }
  }, [isAuthenticated]);

  if (isAuthenticated) {
    return null;
  }

  const onSubmit = (payload: SubmitSelfServiceLoginFlowBody) =>
    flow
      ? newKratosSdk(project)
          .submitSelfServiceLoginFlow(flow.id, payload, sessionToken)
          .then(({ data }) => Promise.resolve(data as SessionContext))
          // Looks like everything worked and we have a session!
          .then((session) => {
            setSession(session);
          })
          .catch(handleFormSubmitError(setFlow, initializeFlow))
      : Promise.resolve();

  return (
    <PageTemplate>
      <View
        style={{
          height: Dimensions.get("window").height,
          width: "100%",
          paddingHorizontal: 20,
          justifyContent: "center",
        }}
      >
        <Card style={{ padding: 20, borderRadius: 10 }}>
          <TextField
            value={email}
            onChange={setEmail}
            style={{ marginBottom: 10 }}
            placeholderText={"Epost"}
          />
          <TextField
            value={password}
            onChange={setPassword}
            password={true}
            placeholderText={"Passord"}
          />
          <Button
          variant="contained"
            onClick={() => {
              const userInput: SubmitSelfServiceLoginFlowWithPasswordMethodBody =
                {
                  identifier: email,
                  method: "password",
                  password: password,
                };
              onSubmit(userInput);
            }}
          >
            <Text style={{fontSize: 18}}>Logg inn</Text>
          </Button>

          <Button
            variant="outlined"
            onClick={() => {
              //@ts-ignore'
              navigation.navigate("signup");
            }}
          >
            <Text style={{color: colors.text_white}}>Opprett ny bruker</Text>
          </Button>

          <Button><Text style={{color: ""}}>Glemt passord?</Text></Button>
        </Card>
      </View>
    </PageTemplate>
  );
}
