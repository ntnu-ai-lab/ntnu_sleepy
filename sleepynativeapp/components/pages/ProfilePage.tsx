import { PageTemplate } from "../material/PageTemplate";
import React, { useContext, useState } from "react";
import { View, Text, ScrollView } from "react-native";
import { relationshipStatus, User, UserEx } from "../../types/Types";
import { gender } from "../../types/Types";
import { AuthContext } from "../../auth/AuthProvider";
import { getTest } from "../../api/userApi";
import { useRecoilState, useRecoilValue } from "recoil";
import { loggedInUser } from "../../state/atoms";
import AsyncStorage from "@react-native-async-storage/async-storage";
import {
  Card,
  Button,
  TextInput,
  Divider,
} from "react-native-paper";
import { colors } from "../../styles/styles";
import SelectDropdown from "react-native-select-dropdown";
import { useNavigation } from "@react-navigation/native";
import { Alert } from "../material/Alert";

export function ProfilePage() {
  const [state, setState] = useState(UserEx);
  const [thisUser] = useRecoilState(loggedInUser);
  const { sessionToken, session, setSession } = useContext(AuthContext);

  const [expanded, setExpanded] = useState(true);
  const navigation = useNavigation();

  const getLocalUser = async () => {
    const user_string = await AsyncStorage.getItem("Local_User");
    return user_string != null ? JSON.parse(user_string) : null;
  };

  const user = getLocalUser();

  const genders = { male: "Mann", female: "Kvinne", other: "Annet" };
  const relationshipStatuses = {
    married: "Gift",
    coliving: "Samboer",
    relationship: "Fast forhold",
    single: "Singel",
  };
  const [gender, setGender] = useState<string>(UserEx.gender);
  const [relationship, setRelationship] = useState<string>(
    UserEx.relationshipStatus
  );
  const [openAlert, setOpenAlert] = useState(false)

  const onEmailChange = (arg: string) =>
    setState((prev) => ({
      ...prev,
      email: arg,
    }));

  const onPasswordChange = (arg: string) =>
    setState((prev) => ({
      ...prev,
      password: arg,
    }));

  const onGenderChange = (arg: string) =>
    setState((prev) => ({
      ...prev,
      gender: arg as gender,
    }));

  const onOccupationChange = (arg: string) =>
    setState((prev) => ({
      ...prev,
      occupation: arg,
    }));

  const onNameChange = (arg: string) =>
    setState((prev) => ({
      ...prev,
      name: arg,
    }));

  const onDateOfBirthChange = (arg: string) =>
    setState((prev) => ({
      ...prev,
      dateOfBirth: arg,
    }));

  const onRelationshipStatusChange = (arg: string) =>
    setState((prev) => ({
      ...prev,
      relationshipStatus: arg as relationshipStatus,
    }));

  function logOut() {
    //@ts-ignore
    setSession({ session_token: undefined, session: undefined });

    const removeUserFromStorage = async () => {
      try {
        await AsyncStorage.removeItem("Local_user");
      } catch (e) {
        console.error;
      }
    };
    removeUserFromStorage();
    setTimeout(() => {
      //@ts-ignore
      navigation.navigate("login");
    }, 100);
  }

  return (
    <PageTemplate>
      <Card style={{ margin: 10 }}>
        <Card.Actions style={{ alignSelf: "center", marginBottom: 10 }}>
          <ScrollView
            style={{ width: "100%" }}
            automaticallyAdjustKeyboardInsets={true}
            alwaysBounceVertical={false}
            endFillColor={colors.primary_dark}
            scrollToOverflowEnabled={true}
          >
            <Divider style={{ margin: 5 }} />
            <TextInput
              editable={false}
              value={state.email}
              onChangeText={onEmailChange}
              placeholder={state.email}
              label="Epostadresse"
              mode="outlined"
            />
            <Divider style={{ margin: 5 }} />
            <TextInput
              value={thisUser ? thisUser.name : "Logg inn på nytt"}
              editable={false}
              label="Navn"
              mode="outlined"
              onChangeText={onNameChange}
            />
            <Divider style={{ margin: 5 }} />
            <TextInput
              editable={false}
              value={state.dateOfBirth}
              onChangeText={onDateOfBirthChange}
              placeholder={state.dateOfBirth}
              label="Fødselsdato"
              mode="outlined"
            />
            <Divider style={{ margin: 5 }} />

            {/* <Select
              value={state.gender}
              onChange={onGenderChange}
              placeholderText={state.gender}
              options={["male", "female", "other"]}
              optionDisplay={(arg: gender) => {
                if (arg === "male") return "Mann";
                else if (arg === "female") return "Kvinne";
                else return "Annet";
              }}
            /> */}

            <SelectDropdown
              disabled={true}
              defaultButtonText={UserEx.gender}
              selectedRowStyle={{ backgroundColor: colors.primary_dark }}
              defaultValue={UserEx.gender}
              buttonStyle={{ width: "100%", borderRadius: 10 }}
              buttonTextStyle={{ textAlign: "left" }}
              dropdownStyle={{ width: "90%", borderRadius: 10 }}
              data={[genders.male, genders.female, genders.other]}
              onSelect={(selectedItem, index) => {
                //console.log(selectedItem, index);
                setGender(selectedItem);
              }}
              buttonTextAfterSelection={(selectedItem, index) => {
                // text represented after item is selected
                // if data array is an array of objects then return selectedItem.property to render after item is selected
                return selectedItem;
              }}
              rowTextForSelection={(item, index) => {
                // text represented for each item in dropdown
                // if data array is an array of objects then return item.property to represent item in dropdown
                return item;
              }}
            />

            <Divider style={{ margin: 5 }} />
            {/*
            <SelectDropdown
              disabled={true}
              defaultButtonText={UserEx.relationshipStatus}
              selectedRowStyle={{ backgroundColor: colors.primary_dark }}
              defaultValue={UserEx.relationshipStatus}
              buttonStyle={{ width: "100%", borderRadius: 10 }}
              buttonTextStyle={{ textAlign: "left" }}
              dropdownStyle={{ width: "90%", borderRadius: 10 }}
              data={[
                relationshipStatuses.married,
                relationshipStatuses.coliving,
                relationshipStatuses.relationship,
                relationshipStatuses.single,
              ]}
              onSelect={(selectedItem, index) => {
                //console.log(selectedItem, index);
                setRelationship(selectedItem);
              }}
              buttonTextAfterSelection={(selectedItem, index) => {
                // text represented after item is selected
                // if data array is an array of objects then return selectedItem.property to render after item is selected
                return selectedItem;
              }}
              rowTextForSelection={(item, index) => {
                // text represented for each item in dropdown
                // if data array is an array of objects then return item.property to represent item in dropdown
                return item;
              }}
            />
            */}
            <Divider style={{ margin: 5 }} />

            {/* <Select
              value={state.relationshipStatus}
              onChange={onRelationshipStatusChange}
              placeholderText={state.relationshipStatus}
              options={["married", "coliving", "relationship", "single"]}
              optionDisplay={(arg: relationshipStatus) => {
                if (arg === "married") return "Gift";
                if (arg === "coliving") return "Samboer";
                if (arg === "relationship") return "Fast forhold";
                return "Singel";
              }}
            /> */}
            <TextInput
              editable={false}
              value={state.occupation}
              onChangeText={onOccupationChange}
              placeholder={state.occupation}
              label="Yrke"
              mode="outlined"
            />
          </ScrollView>
        </Card.Actions>
        <Button
          mode="text"
          onPress={() => {
            logOut();
          }}
        >
          <Text>Logg ut</Text>
        </Button>
        <Button
          mode="text"
          onPress={() => {
            if (sessionToken && session)
              getTest("d10e1d18-db25-4e47-ab19-efad53248123").then(r => {
                if (r === 0) {
                  setOpenAlert(true)
                }
              });
          }}
        >
          <Text>Test</Text>
        </Button>
      </Card>

      <View style={{ height: 80, width: "100%" }} />
      <Alert type={"alert"} content={"Klarte ikke å hente test, dette kan skyldes mange på internett, eller mangel på kompetanse fra vår side"} open={openAlert} setOpen={setOpenAlert}/>
    </PageTemplate>
  );
}
