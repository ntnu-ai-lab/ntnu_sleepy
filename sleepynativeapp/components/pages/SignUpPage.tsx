import {
  SelfServiceRegistrationFlow,
  SubmitSelfServiceRegistrationFlowBody,
  SubmitSelfServiceRegistrationFlowWithPasswordMethodBody,
} from "@ory/kratos-client";
import { useFocusEffect, useNavigation } from "@react-navigation/native";
import React, { useContext, useState } from "react";
import { View, Text, ScrollView, Dimensions } from "react-native";
import { createUser } from "../../api/userApi";
import { AuthContext } from "../../auth/AuthProvider";
import { handleFormSubmitError } from "../../auth/form";
import { ProjectContext } from "../../auth/ProjectProvider";
import { newKratosSdk } from "../../auth/Sdk";
import { User, gender, relationshipStatus } from "../../types/Types";
import { PageTemplate } from "../material/PageTemplate";
import { colors } from "../../styles/styles";
import { Button } from "../material/Button";
import { Card } from "../material/Card";
import { TextField } from "../material/TextField";
import { Select } from "../material/Select";
import { IconButton } from "../material/IconButton";
import ArrowBack from "../../assets/arrowBack.svg"

export function SignupPage() {
  const [email, setEmail] = useState<string>("");
  const [name, setName] = useState<string>("");
  const [password, setPassword] = useState<string>("");
  const [password2, setPassword2] = useState<string>("");
  const [dateOfBirth, setDateOfBirth] = useState<string>("");
  const [occupation, setOccupation] = useState<string>("");

  const navigation = useNavigation();

  const genders = {
    male: "Mann",
    female: "Kvinne",
    other: "Annet",
    undefined: "-",
  };
  const relationshipStatuses = {
    married: "Gift",
    coliving: "Samboer",
    relationship: "Fast forhold",
    single: "Singel",
    undefined: "-",
  };

  const [gender, setGender] = useState<gender>("undefined");
  const [relationship, setRelationship] =
    useState<relationshipStatus>("undefined");

  const cardHeight = Dimensions.get("window").height * 0.8 

  function passwordMatch() {
    if (!password2) return true;
    else if (password === password2) return true;
    return false;
  }

  const [flow, setConfig] = useState<SelfServiceRegistrationFlow | undefined>(
    undefined
  );
  const { project } = useContext(ProjectContext);
  const { setSession, isAuthenticated, sessionToken, session } =
    useContext(AuthContext);

  const initializeFlow = () => {
    newKratosSdk(project)
      .initializeSelfServiceRegistrationFlowWithoutBrowser()
      // The flow was initialized successfully, let's set the form data:
      .then(({ data: flow }) => {
        setConfig(flow);
      })
      .catch(console.error);
  };

  // When the component is mounted, we initialize a new use login flow:
  useFocusEffect(
    React.useCallback(() => {
      initializeFlow();

      return () => {
        setConfig(undefined);
      };
    }, [project])
  );

  // This will update the registration flow with the user provided input:
  const onSubmit = (
    payload: SubmitSelfServiceRegistrationFlowBody
  ): Promise<void> =>
    flow
      ? newKratosSdk(project)
          .submitSelfServiceRegistrationFlow(flow.id, payload)
          .then(({ data }) => {
            // ORY Kratos can be configured in such a way that it requires a login after
            // registration. You could handle that case by navigating to the Login screen
            // but for simplicity we'll just print an error here:
            if (!data.session_token || !data.session) {
              const err = new Error(
                "It looks like you configured ORY Kratos to not issue a session automatically after registration. This edge-case is currently not supported in this example app. You can find more information on enabling this feature here: https://www.ory.sh/kratos/docs/next/self-service/flows/user-registration#successful-registration"
              );
              return Promise.reject(err);
            }

            // Looks like we got a session!
            return Promise.resolve({
              session: data.session,
              session_token: data.session_token,
            });
          })
          // Let's log the user in!
          .then((s) => {
            setSession(s);
            const user: User = {
              name: name,
              email: email,
              username: email,
              dateOfBirth: dateOfBirth,
              gender: gender,
              occupation: occupation,
              relationshipStatus: relationship,
            };
            createUser(user, s.session.identity.id); //@ts-ignore
            navigation.navigate("home");
          })
          .catch(
            handleFormSubmitError<SelfServiceRegistrationFlow | undefined>(
              setConfig,
              initializeFlow
            )
          )
      : Promise.resolve();

  return (
    <PageTemplate style={{ paddingTop: 60, paddingBottom: 100 }}>
      <View style={{position: "absolute", top: 60, left: 10, zIndex: 10000}}>
        <IconButton onClick={navigation.goBack}>
          <ArrowBack/>
        </IconButton>
      </View>
      <View style={{ marginVertical: 10 }}>
        <Text style={{ fontSize: 24, alignSelf: "center" }}>
          Registrer ny bruker
        </Text>
      </View>
      <Card style={{ alignSelf: "center", width: "95%", height: cardHeight }}>
        <View>
          <ScrollView style={{height: cardHeight - 85}}>
            <TextField
              value={name}
              onChange={setName}
              placeholderText="Fullt navn"
              style={{ marginBottom: 10 }}
            />
            <TextField
              value={email}
              onChange={setEmail}
              placeholderText="E-postadresse"
              style={{ marginBottom: 10 }}
            />
            <TextField
              value={password}
              onChange={setPassword}
              placeholderText="Passord"
              password={true}
              style={{ marginBottom: 10 }}
            />
            <TextField
              value={password2}
              onChange={setPassword2}
              placeholderText="Oppgi passord på nytt"
              password={true}
              error={!passwordMatch()}
              style={{ marginBottom: 10 }}
            />
            <View>
              <Text style={{fontSize: 12, color: colors.text_white, textTransform: "uppercase"}}>Sivilstatus</Text>
            </View>
            <View>
              <Select
                options={[
                  relationshipStatuses.married,
                  relationshipStatuses.coliving,
                  relationshipStatuses.relationship,
                  relationshipStatuses.single,
                  relationshipStatuses.undefined,
                ]} 
                optionDisplay={(o) => o}
                onChange={setRelationship}
                />
            </View>
            <View>
              <Text style={{fontSize: 12, color: colors.text_white, textTransform: "uppercase"}}>Kjønn</Text>
            </View>
            <View>
              <Select
                options={[
                  genders.male,
                  genders.female,
                  genders.other,
                  genders.undefined,
                ]}
                optionDisplay={(o) => o}
                onChange={setGender}
              />
            </View>
            <TextField
              value={dateOfBirth}
              onChange={setDateOfBirth}
              style={{ marginBottom: 10 }}
              placeholderText="Fødselsdato på format: ddmmåååå"
            />
            <TextField
              value={occupation}
              onChange={setOccupation}
              style={{ marginBottom: 10 }}
              placeholderText="Yrke"
            />
          </ScrollView>
        </View>
        <View style={{height: 80}}>
        <Button
          variant="contained"
          onClick={() => {
            const userInput: SubmitSelfServiceRegistrationFlowWithPasswordMethodBody =
              {
                method: "password",
                password: password,
                traits: { email: email },
              };
            onSubmit(userInput);
          }}
        >
          <Text style={{ fontSize: 18 }}>Registrer</Text>
        </Button>
        </View>
      </Card>
    </PageTemplate>
  );
}
