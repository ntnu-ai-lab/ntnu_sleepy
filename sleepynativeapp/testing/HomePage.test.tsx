import { HomePage } from '../components/pages/HomePage';
import renderer from 'react-test-renderer';

describe("Test HomePage: ", () => {

    it("Renders correctly", async () => {
        const component = renderer.create(<HomePage />).toJSON()
        expect(component).toMatchSnapshot();
    })
})