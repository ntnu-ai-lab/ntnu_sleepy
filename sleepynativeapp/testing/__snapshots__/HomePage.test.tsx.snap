// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HomePage:  Renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#A5C9CA",
        "height": "100%",
        "paddingTop": 0,
        "width": "100%",
      },
      undefined,
    ]
  }
>
  <View>
    <Text>
      This is the homepage
    </Text>
  </View>
</View>
`;
